src/game/requester.c

```diff
@@ -174,34 +189,28 @@ void Requester_SetHeading(REQUEST_INFO *req, const char *string)
     Text_Remove(req->heading);
     req->heading = NULL;
 
-    if (string) {
-        req->heading_text = string; // unsafe
-    } else {
-        req->heading_text = NULL;
-    }
+    size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+    snprintf(req->heading_text, out_size, "%s", string);
 }
 
 void Requester_ChangeItem(
     REQUEST_INFO *req, int32_t idx, const char *string, uint16_t flag)
 {
-    if (string) {
-        strcpy(&req->item_texts[idx * req->item_text_len], string);
+    if (idx > 0 && idx < req->max_items && string) {
+        size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+        snprintf(req->item_texts[idx], out_size, "%s", string);
         req->item_flags[idx] = flag;
-    } else {
-        req->item_flags[idx] = 0;
     }
 }
 
 void Requester_AddItem(REQUEST_INFO *req, const char *string, uint16_t flag)
 {
-    if (string) {
-        strcpy(&req->item_texts[req->items * req->item_text_len], string);
-        req->item_flags[req->items] = flag;
-    } else {
-        req->item_flags[req->items] = 0;
+    if (req->items_used < req->max_items && string) {
+        size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+        snprintf(req->item_texts[req->items_used], out_size, "%s", string);
```

Comment:
This is the place where I'd do the string allocation. To make this function even more useful, I'd make it accept a `format` string and a variable number of arguments using the `va_start()` family of functions (see how `Log_Message` is implemented). This would allow the callers to do `Requester_AddItem(req, flag, "%s", level_title)` without having to care about memory allocation at all. Naturally this idiom would also need to be applied to `Requester_ChangeItem` that would have to reallocate the item text buffer.

Reply:
`Requester_ChangeItem ` I noticed is never used. Do we want to keep it?